---
title: "AD Genetics Descriptives"
output: html_document
date: "`r format(Sys.time(), '%d %B, %Y')`"
author:
    - "Shea Andrews"
params:
    rmd: "report.Rmd"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidyr)
library(purrr)
library(janitor)
library(tibble)
library(glue)
library(stringr)
library(readr)
library(forcats)
library(glue)
library(readxl)
# library(DT)
`%nin%` <- negate(`%in%`)

```


```{r import, echo = F, include=F}
path_ad_loci <- "results/adgwas_loci.csv"
path_adgwas_meta <- "intermediate/gwas_metadata.csv"
path_supp_wightman <- "resources/Wightman2020/41588_2021_921_MOESM4_ESM.xlsx"
path_supp_bellenguez <- "resources/Bellenguez2022/41588_2022_1024_MOESM4_ESM.xlsx"
path_clean_bellenguez <- "intermediate/bellenguez_clean_loci.csv"

# Export
path_table_s1 <- "docs/TableS1.csv"

if (exists("snakemake")) {
  path_supp_bellenguez <- snakemake@input[["supp_bellenguez"]]
  path_clean_bellenguez <- snakemake@input[["bellenguez_clean"]]
  path_supp_wightman <- snakemake@input[["supp_wightman"]]
  path_ad_loci <- snakemake@input[["ad_loci"]]
  path_adgwas_meta <- snakemake@input[["ad_gwas_meta"]]
  path_table_s1 <- snakemake@params[["TableS1"]]
}

raw_ad_loci <- read_csv(path_ad_loci)
adgwas_meta <- read_csv(path_adgwas_meta)
```


```{r, echo = F}
ad_loci <- raw_ad_loci %>%
  filter(study %nin% c("Jonsson", "Reiman")) %>%
  select(locus, cytoband, locus_ld, Gene = gencode_gene, SNP, CHR, BP,
         A1, A2, FRQ, gnomad_af, OR, P, study) %>%
  mutate(
    FRQ = signif(FRQ, 3),
    OR = signif(OR, 3),
  )

write_csv(ad_loci, path_table_s1)
```

```{r, echo=F, include=F}
## Distinct loci and independent across all studies
n_distinct(ad_loci$locus)
n_distinct(ad_loci$locus_ld)

## Old loci prior to Bellenguez and Wightman
known_loci <- ad_loci %>%
  filter(study %nin% c("Bellenguez", "Wightman"))

count(known_loci, study)
n_distinct(known_loci$locus)

## 42 novel loci discovered by Bellenguez and Whightman
novel_loci <- ad_loci %>%
  filter(study %in% c("Bellenguez", "Wightman")) %>% 
  anti_join(known_loci, by = "locus") %>%
  distinct(locus, .keep_all = TRUE) %>%
  nrow(.)

novel_loci_study <- ad_loci %>%
  filter(study %in% c("Bellenguez", "Wightman")) %>% 
  anti_join(known_loci, by = "locus") %>%
  group_by(study) %>%
  distinct(locus, .keep_all = TRUE) %>%
  ungroup()
  nrow()

count(novel_loci_study, study)

# Number of loci in B & W
n_loci <- ad_loci %>%
  filter(study %in% c("Bellenguez", "Wightman")) %>%
  group_by(locus)

n_distinct(n_loci$locus)

# Number of loci per study
n_loci_study <- ad_loci %>%
  group_by(study)  %>%
  distinct(locus) %>%
  ungroup() %>%
  count(study)

n_loci_study

# Number of independent variants
n_variants <- ad_loci %>%
  filter(study %in% c("Bellenguez", "Wightman")) %>%
  group_by(locus_ld)

n_distinct(n_variants$locus_ld)

# Number of independent variants per study
n_variants_study <- ad_loci %>%
  group_by(study)  %>%
  distinct(locus_ld) %>%
  ungroup() %>%
  count(study)

n_variants_study

```

## Genetic loci associated with Alzheimerâ€™s disease

Wightman et al expanded the sample size of their GWAS to `r filter(adgwas_meta, study == 'Wightman') %>% pull(n)` 
people (`r `filter(adgwas_meta, study == 'Wightman') %>% pull(n_ca)` cases, 
`r NULL` proxy-cases, `r filter(adgwas_meta, study == 'Wightman') %>% pull(n_cn)` controls/proxy-controls). 


Bellenguez et al expanded their earlier GWAS to `r filter(adgwas_meta, study == 'Bellenguez') %>% pull(n)` 
people (`r filter(adgwas_meta, study == 'Bellenguez') %>% pull(n_ca)` cases, 
`r filter(adgwas_meta, study == 'Bellenguez') %>% pull(n_pca)` proxy cases, and 
`r filter(adgwas_meta, study == 'Bellenguez') %>% pull(n_cn)` controls/proxy-controls)


Together, Wightman et al and Bellenguez et al identified `r n_distinct(n_variants$locus_ld)` independent variants 
across `r n_distinct(n_loci$locus)` loci 
(`r filter(n_loci_study, study == "Wightman") %>% pull(n)` and `r filter(n_loci_study, study == "Bellenguez") %>% pull(n)` 
respectively) associated with AD, 34 of which were identified by both studies and `r nrow(novel_loci)` which are novel loci.


With GWAS collectively discovering `r n_distinct(ad_loci$locus)` loci 
(`r n_distinct(ad_loci$locus_ld)` independent variants) associated with AD/dementia 
in European ancestry populations (Figure 2), how many more novel loci remain to be discovered?


```{r, echo = F}
DT::datatable(ad_loci, rownames = FALSE, filter = "top")
```


```{r, echo=F, include=F}
# Which loci from Bellenguez are getting collapsed?
clean_bellenguez <- read_csv(path_clean_bellenguez)

bell_org <- clean_bellenguez %>%
  filter(SNP != "rs429358")

n_distinct(bell_org$LOCUS)

bell_munge <- raw_ad_loci %>% 
  filter(study == "Bellenguez") %>%
  select(locus, locus_ld, SNP, CHR, BP, A1, A2, GENE, FRQ, OR, P) %>%
  filter(SNP != "rs429358") 

n_distinct(bell_munge$locus_ld)
n_distinct(bell_munge$locus)

bell_comb <- full_join(
  select(bell_org, SNP, LOCUS), bell_munge, 
  by = "SNP"
)

bell_comb %>%
  janitor::get_dupes(locus) %>%
  print(n = Inf)

## Collapsed loci CLU/PTK2B, UMAD1/ICA1, ADAMTS1/APP (42), ABCA7/KLF16 (37), APH1B/SNX1 (28)

filter(bell_comb, GENE %in% c(
  "CLU", "PTK2B", "UMAD1", "ICA1", "ADAMTS1", "APP", "ABCA7", "KLF16", "APH1B", "SNX1"
))

```


```{r, include=F, echo=F}
## Add priortized genes to bellenguez
bst20_raw <- readxl::read_xlsx(path_supp_bellenguez, sheet = 21, skip = 2) %>%
  janitor::clean_names()

bst20 <- bst20_raw %>%
  select(locus, gene, gene_prioritization_tier) %>%
  filter(gene_prioritization_tier %in% c("Tier 1", "Tier 2")) %>%
  mutate(locus_n = str_extract(locus, '(?<=\\().*[:digit:](?=\\))')) %>%
  rename(locus_gene = locus)

bell_comb_gene <- bell_comb %>%
  left_join(bst20, by = c("LOCUS" = "locus_n")) 

## Number of distinct loci with a Teir 1 gene  
bell_comb_gene %>%
  filter(nchar(LOCUS) <= 2) %>%
  filter(gene_prioritization_tier == "Tier 1") %>%
  distinct(locus)

## Number of genes in each teir
### From all newly identified loci, this gene prioritization yielded 31 tier 1 genes and 24 tier 2. Among the 31 tier 1 genes, we observed that 25 of these genes were the only prioritized gene in their respective locus. For the remaining 6 tier 1 genes, we also found tier 2 genes in their respective locus. We also identified five loci containing several tier 2 prioritized genes... In five loci, our prioritization score did not identify sufficient molecular evidence to prioritize genes with exception of being the nearest gene (L10, L12, L13, L14 and L32). 

bell_comb_gene %>%
  # filter(LOCUS %in% c("10", "12", "13", "14", "32"))
  filter(nchar(LOCUS) <= 2) %>%
  # mutate(gene_prioritization_tier = ifelse(LOCUS %in% c("10", "12", "13", "14", "32"), NA, gene_prioritization_tier)) %>%
  filter(!is.na(gene_prioritization_tier)) %>%
  distinct(gene, .keep_all = T) %>%
  count(gene_prioritization_tier)

## Number of loci with each tier category
bell_comb_gene %>%
  filter(nchar(LOCUS) <= 2) %>%
  # mutate(gene_prioritization_tier = ifelse(LOCUS %in% c("10", "12", "13", "14", "32"), NA, gene_prioritization_tier)) %>%
  filter(!is.na(gene_prioritization_tier)) %>%
  distinct(gene, .keep_all = T) %>%
  group_by(LOCUS) %>%
  count(gene_prioritization_tier) %>% 
  pivot_wider(names_from = gene_prioritization_tier, values_from = n) %>%
  janitor::clean_names() %>%
  ungroup() %>%
  mutate(
    cat = case_when(
      tier_1 == 1 & is.na(tier_2) ~ "Tier 1 only", 
      tier_1 == 1 & tier_2 == 1 ~ "Tier 1 + 2", 
      tier_1 == 1 & tier_2 > 1 ~ "Tier 1 + multi Tier 2", 
      is.na(tier_1) & tier_2 == 1 ~ "Tier 2 only", 
      is.na(tier_1) & tier_2 > 1 ~ "mutlti Tier 2 ", 
    )
  ) %>%
  count(cat)

## Our locus definition
bell_comb_gene %>%
  filter(nchar(LOCUS) <= 2) %>%
  # mutate(gene_prioritization_tier = ifelse(LOCUS %in% c("10", "12", "13", "14", "32"), NA, gene_prioritization_tier)) %>%
  filter(!is.na(gene_prioritization_tier)) %>%
  distinct(gene, .keep_all = T) %>%
  group_by(locus) %>%
  count(gene_prioritization_tier) %>% 
  pivot_wider(names_from = gene_prioritization_tier, values_from = n) %>%
  janitor::clean_names() %>%
  ungroup() %>%
  mutate(
    cat = case_when(
      tier_1 == 1 & is.na(tier_2) ~ "Tier 1 only", 
      tier_1 == 1 & tier_2 == 1 ~ "Tier 1 + 2", 
      tier_1 == 1 & tier_2 > 1 ~ "Tier 1 + multi Tier 2", 
      is.na(tier_1) & tier_2 == 1 ~ "Tier 2 only", 
      is.na(tier_1) & tier_2 > 1 ~ "mutlti Tier 2 ", 
    )
  ) %>%
  count(cat)

wst11 <- readxl::read_xlsx(path_supp_wightman, sheet = 11, skip = 1) %>%
  janitor::clean_names()

bst20 %>%
   filter(gene_prioritization_tier == "Tier 1") %>%
   print(n = Inf)



```



































